numpy/core/einsumfunc.py

```diff
@@ -0,0 +1,597 @@
+from numpy.core.multiarray import einsum as c_einsum
+from numpy.core.numeric import asarray, asanyarray, result_type
+
+__all__ = ['einsum']
+
+
+einsum_symbols = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
+einsum_symbols_set = set(einsum_symbols)
+
+def _compute_size_by_dict(indices, idx_dict):
+    """
+    Computes the product of the elements in indices based on the dictionary
+    idx_dict.
+
+    Parameters
+    ----------
+    indices : iterable
+        Indices to base the product on.
+    idx_dict : dictionary
+        Dictionary of index sizes
+
+    Returns
+    -------
+    ret : int
+        The resulting product.
+    """
+    ret = 1
+    for i in indices:
+        ret *= idx_dict[i]
+    return ret
+
+
+def _find_contraction(positions, input_sets, output_set):
```

Comment:
This is just my personal preference style. I like what @mrocklin does in dask, e.g., https://github.com/blaze/dask/blob/c4df983f03eeac02b5b99ba0628b2ead84b42a17/dask/array/core.py#L1204

In my opinion, examples can actually be better than full numpydoc style documentation of arguments. I was scratching my head a little bit to understand the difference between the "resulting" and "current" contraction :).